pkg/controller/drain.go

```diff
@@ -373,6 +387,236 @@ func (o *DrainOptions) evictPods(pods []api.Pod, policyGroupVersion string, getP
+				select {
+				case <-detachCh:
+					glog.V(3).Info("Detached pv for pod: ", pod.Name)
+					returnCh <- nil
+				case <-time.After(o.PvDetachTimeout - time.Now().Sub(startTime)):
+					glog.Warningf("Timout occured when detaching pv for pod: %v", pod.Name)
+					returnCh <- nil
+				}
+			case <-time.After(o.PvDetachTimeout):
+				returnCh <- fmt.Errorf("Timout when evicting pod: %v", pod)
+			}
+		}
+		if len(retryPods) == 0 {
+			break
+		}
+		glog.V(4).Info("Eviction for some pods will be retried after 5 seconds.")
+		pods = retryPods
+		time.Sleep(5 * time.Second)
+	}
+	return
+}
+
+func (o *DrainOptions) getPvs(pod corev1.Pod) []string {
+	pvs := []string{}
+	for _, vol := range pod.Spec.Volumes {
+		if vol.PersistentVolumeClaim != nil {
+			pvc, err := o.client.CoreV1().PersistentVolumeClaims(pod.Namespace).Get(vol.PersistentVolumeClaim.ClaimName, metav1.GetOptions{})
+			if err != nil {
+				// If error is not nil, then skip this volume
+				continue
+			}
+			pvs = append(pvs, pvc.Spec.VolumeName)
+		}
+	}
+	return pvs
+}
+
+func (o *DrainOptions) waitForDetach(pvNames []string, nodeName string, detachCh chan bool) {
+	if pvNames == nil || len(pvNames) == 0 || nodeName == "" {
+		// If pv or node name is not available, nothing to do. Just log this as warning
+		glog.Warningf("Node name: %v, list of PVs: %v", nodeName, pvNames)
+		detachCh <- false
+		return
+	}
+	volumeIDs, err := o.getVolIDsFromDriver(pvNames)
+	if err != nil {
+		glog.Errorf("Error when getting volume IDs from the driver. Err: %v", err)
+		detachCh <- false
+		return
+	}
+	glog.V(4).Info("Volume names from driver: ", volumeIDs)
+
+	found := true
+	for found {
+		found = false
+		node, err := o.client.Core().Nodes().Get(nodeName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			glog.V(4).Info("Node not found: ", nodeName)
+			detachCh <- true
+			return
+		} else if err != nil {
+			glog.Errorf("Error getting details for node: %v. Err: %v", nodeName, err)
+			detachCh <- false
+			return
+		}
+
+		glog.V(4).Info("Attached volumes: ", node.Status.VolumesAttached)
+		attachedVols := node.Status.VolumesAttached
+		if len(attachedVols) == 0 {
+			glog.V(4).Info("No volumes attached to the node")
+			break
+		}
+
+		for _, volumeID := range volumeIDs {
+			for _, attachedVol := range attachedVols {
+				found, _ = regexp.MatchString(volumeID, string(attachedVol.Name))
+				if found {
+					glog.V(4).Info("Found volume: ", volumeID, " attached to the node")
+					break
+				}
+			}
+			if found {
+				glog.V(4).Info("Will recheck in 5 seconds")
+				time.Sleep(5 * time.Second)
+				break
+			}
+		}
+	}
+	glog.V(4).Info("Detached PVs: ", pvNames)
+	detachCh <- true
+}
+
+func (o *DrainOptions) getVolIDsFromDriver(pvNames []string) ([]string, error) {
+	pvSpecs := []corev1.PersistentVolumeSpec{}
+	for _, pvName := range pvNames {
+		pv, err := o.client.CoreV1().PersistentVolumes().Get(pvName, metav1.GetOptions{})
```

Comment:
If we miss out on the volume due to temp-issues of APIServer, do we skip waiting for that volume? I dont have inclination towards any approach, just pointing out the observed behaviour. :-P

Reply:
Please make the driver call in [`getPVs`](https://github.com/gardener/machine-controller-manager/pull/275/files#diff-face99aebd3cd79577003f7df747c863R502) and skip this func altogether.